consensus/core/src/commit_syncer.rs

```diff
@@ -0,0 +1,557 @@
+                    return Err(ConsensusError::UnexpectedStartCommit {
+                        peer,
+                        start,
+                        commit: Box::new(commit),
+                    });
+                }
+            } else {
+                // Verify next commit increments index and references the previous digest.
+                let (last_commit_digest, last_commit): &(CommitDigest, Commit) =
+                    commits.last().unwrap();
+                if commit.index() != last_commit.index() + 1
+                    || &commit.previous_digest() != last_commit_digest
+                {
+                    return Err(ConsensusError::UnexpectedCommitSequence {
+                        peer,
+                        prev_commit: Box::new(last_commit.clone()),
+                        curr_commit: Box::new(commit),
+                    });
+                }
+            }
+            // Do not process more commits past the end index.
+            if commit.index() > end {
+                break;
+            }
+            commits.push((digest, commit));
+        }
+        let Some((end_commit_digest, end_commit)) = commits.last() else {
+            return Err(ConsensusError::NoCommitReceived { peer });
+        };
+
+        // Parse and verify blocks. Then accumulate votes on the end commit.
+        let end_commit_ref = CommitRef::new(end_commit.index(), *end_commit_digest);
+        let mut stake_aggregator = StakeAggregator::<QuorumThreshold>::new();
+        for serialized in serialized_blocks {
+            let block: SignedBlock =
+                bcs::from_bytes(&serialized).map_err(ConsensusError::MalformedBlock)?;
+            // The block signature needs to be verified.
+            self.block_verifier.verify(&block)?;
+            for vote in block.commit_votes() {
+                if *vote == end_commit_ref {
+                    stake_aggregator.add(block.author(), &self.context.committee);
+                }
+            }
+        }
+
+        // Check if the end commit has enough votes.
+        if !stake_aggregator.reached_threshold(&self.context.committee) {
+            return Err(ConsensusError::NotEnoughCommitVotes {
+                stake: stake_aggregator.stake(),
+                peer,
+                commit: Box::new(end_commit.clone()),
+            });
+        }
+
+        Ok(commits
+            .into_iter()
+            .zip(serialized_commits)
+            .map(|((_d, c), s)| TrustedCommit::new_trusted(c, s))
+            .collect())
+    }
+}
+
+struct FetchState {
+    // Tuple of the first time an authority is available to be fetched, authority index and
+    // previous consequtive failures count.
+    available_authorities: BTreeSet<(Instant, AuthorityIndex, u32)>,
+}
+
+impl FetchState {
+    fn new(context: &Context) -> Self {
+        // Randomize the initial order of authorities.
+        let mut available_authorities: Vec<_> = context
```

Comment:
Add a todo to use scores here? If you agree we can use the scores to fetch from "good" hosts instead of all available hosts. We can do this in live syncer as well

Reply:
Added to the comment above `FetchState.available_authorities`. This component needs to balance load among as many peers as possible, while avoid unhealthy peers.